### PR TITLE
Make alias input_type protected for join_node

### PR DESCRIPTION
**Pull request by dbiswas2808**
Currently the public aliases exposed by `flow_graph::join_node` includes `input_type`. This is a bit confusing since `join_node` has multiple input ports and I believe other nodes(with multiple inputs) don't have the same issue. I have been leveraging substitution failure with `input_ports_type` and `input_type` and similarly for `output_ports_type` and `output_type` for implementing a higher level interface for flow_graph. AFAIT only `join_node` breaks this distinction. Also, having `join_node` expose `input_type` as a public alias could be confusing to users.

### CODE REVIEW

File: include/oneapi/tbb/detail/_flow_graph_join_impl.h
Comment (kboyarinov) [inline include/oneapi/tbb/detail/_flow_graph_join_impl.h:1029]: I have maid a quick check and it seems like this alias it is not needed outside the `join_node_FE` class - as well as other public aliases defined there. It seems like it would be better to make `input_type` and other aliases in `join_node_FE` specializations private. @aleksei-fedotov, what do you think?

Comment (akukanov) [reply include/oneapi/tbb/detail/_flow_graph_join_impl.h:1029]: For what it`s worth. a public class like `join_node` definitely should hide all internal interfaces inherited from the implementation.

**Comment by kboyarinov** [comment]
@dbiswas2808, could you please remove the line related to another PR from this patch and it would be approved and merged.
Thank you

File: include/oneapi/tbb/detail/_flow_graph_join_impl.h
Comment (aleksei-fedotov) [reply include/oneapi/tbb/detail/_flow_graph_join_impl.h]: > I have maid a quick check and it seems like this alias it is not needed outside the `join_node_FE` class - as well as other public aliases defined there. It seems like it would be better to make `input_type` and other aliases in `join_node_FE` specializations private. @aleksei-fedotov, what do you think?

Sorry for a late reply. 
The idea behind input and output type is to make node interfaces common so that users can reference the corresponding type easily by writing something like: `<node_type>::input_type` or `<node_type>::output_type`.
There are exceptions for multi-input/output node types (such as `join_node`, `split_node`). 
Yes, internal interfaces should be hidden. So, the changes are correct (except for the line with static assert).

**Comment by dbiswas2808** [comment]
> @dbiswas2808, could you please remove the line related to another PR from this patch and it would be approved and merged.
> Thank you

Done!

**Review by kboyarinov — DISMISSED** [review]
LGTM

**Comment by dbiswas2808** [comment]
@kboyarinov 
```
    template<typename JP, typename InputTuple, typename OutputTuple>
    class join_node_base : public graph_node, public join_node_FE<JP, InputTuple, OutputTuple>,
                           public sender<OutputTuple> {
    protected:
        using graph_node::my_graph;
    public:
        typedef OutputTuple output_type;
```
I think making the aliases protected might fix the failures. If you look at the above snippet the `OutputTuple` typedef  uses the `OutputType` from `join_node_FE`. What do you think?

**Comment by kboyarinov** [comment]
@dbiswas2808, as I understand the testing failure, the problem is the following ( I have simplified a code a bit):

```
template <typename Tuple>
class join_node_FE {
    static const int N = std::tuple_size<Tuple>::value;
};

template <int N, typename Tuple>
class unfolded_join_node : public join_node_FE<Tuple> {
public:
    typedef typename wrap_tuple_elements<N, Tuple>::type input_ports_type; // problem is here
};
```

Visual Studio compiler for some reason interprets the usage of `N` on the last line as an access to private field of the base class, not the local non-type template parameter `N`.
Unfortunately, I failed to reproduce this on my local machine, I would try on other machines to figure out how this can be fixed

**Comment by dbiswas2808** [comment]
> @dbiswas2808, as I understand the testing failure, the problem is the following ( I have simplified a code a bit):
> 
> ```
> template <typename Tuple>
> class join_node_FE {
>     static const int N = std::tuple_size<Tuple>::value;
> };
> 
> template <int N, typename Tuple>
> class unfolded_join_node : public join_node_FE<Tuple> {
> public:
>     typedef typename wrap_tuple_elements<N, Tuple>::type input_ports_type; // problem is here
> };
> ```
> 
> Visual Studio compiler for some reason interprets the usage of `N` on the last line as an access to private field of the base class, not the local non-type template parameter `N`. Unfortunately, I failed to reproduce this on my local machine, I would try on other machines to figure out how this can be fixed

@kboyarinov That is strange indeed I haven't seen that myself before. Any updates on how to fix this or get around this issue?

File: include/oneapi/tbb/detail/_flow_graph_join_impl.h
Comment (BenFrantzDale) [inline include/oneapi/tbb/detail/_flow_graph_join_impl.h:671]: Should there be a comment here saying why these are private?

Comment (kboyarinov) [reply include/oneapi/tbb/detail/_flow_graph_join_impl.h:671]: As discussed [above](https://github.com/oneapi-src/oneTBB/pull/868#discussion_r950049713), this parameter is part of the internal implementation details of `join_node_FE` so it should not be part of `oneapi::tbb::join_node` public API. 
Do you think we really need an extra comment in the code for this?

**Comment by kboyarinov** [comment]
@dbiswas2808, as a simplest way to workaround a VS issue, I would propose to rename the template parameter of `unfolded_join_node` from `N` to something different. I guess it should fix the problem.

File: include/oneapi/tbb/detail/_flow_graph_join_impl.h
Comment (BenFrantzDale) [reply include/oneapi/tbb/detail/_flow_graph_join_impl.h:671]: 🤷  I'm just trying to consider a future reader looking here without much context thinking "Why are `N`, `output_type`, `input_type`, and `base_node_type` private?" I'm thinking that would be clarified if they saw `// These parameters are private because they are part of the internal implementation details of join_node_FE: After all, a join node doesn't have a single input port.`

Just a thought.

**Comment by dbiswas2808** [comment]
> @dbiswas2808, as a simplest way to workaround a VS issue, I would propose to rename the template parameter of `unfolded_join_node` from `N` to something different. I guess it should fix the problem.

@kboyarinov I just Made the suggested change. Hopefully it compiles fine on MSVC.

**Comment by dbiswas2808** [comment]
@kboyarinov I've fixed the msvc compile issue as you suggested in the last commit. LMK if it looks right.

**Review by kboyarinov — APPROVED** [review]

**Comment by dbiswas2808** [comment]
@kboyarinov I fixed the issue I think. Seems to build on my MacOS system

**Review by kboyarinov — DISMISSED** [review]
Thanks for the contribution! LGTM!

**Comment by kboyarinov** [comment]
@dbiswas2808, one last request from my side. Could you please add a sign-off string to the commit message? See [Contribution.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests:~:text=No%20anonymous%20contributions%20are%20accepted.%20The%20name%20in%20the%20commit%20message%20Signed%2Doff%2Dby%20line%20and%20your%20email%20must%20match%20the%20change%20authorship%20information.%20Make%20sure%20your%20.gitconfig%20is%20set%20up%20correctly%20so%20you%20can%20use%20git%20commit%20%2Ds%20for%20signing%20your%20patches%3A) for details. After that, the PR can be merged to master.
Thank you once again for your contribution.

**Comment by dbiswas2808** [comment]
@kboyarinov I just signed off all the commits. Will need your approval for the pr-builder to start I think

**Review by kboyarinov — APPROVED** [review]

**Pull request merged**
